### ISSUE TITLE
V1 Design

### ISSUE THREAD

**Issue by brillout** [brillout/vite-plugin-ssr#578]
The `vite-plugin-ssr@1.0.0` design in a nutshell:

```ts
// /pages/product/+config.ts

import type { Config } from 'vite-plugin-ssr'

export default {
  Page: './Page.vue',
  route: '/product/@id',
  onBeforeRender: './onBeforeRender.ts',
  onPrerender: './onPrerender.ts',
  prerender: true,
  ssr: false // render as SPA
} satisfies Config // `satisfies` is a new TypeScript 4.9 operator
```

I.e. `+config.js` replaces `.page.js`, `.page.server.js`, `.page.client.js`, and `.page.route.js`.

It also replaces `_default.page.js` as well as VPS configurations set in `vite.config.js`:

```ts
// /pages/+config.ts

import type { Config } from 'vite-plugin-ssr'

// Applies as default to all pages
export default {
  onRenderHtml: './config/onRenderHtml.tsx',
  onRenderClient: './config/onRenderClient.tsx',
  prerender: {
    partial: true // Currently lives in vite.config.js
  },
  ssr: true
} satisfies Config
```

VPS's architecture stays the same and therefore migration is relatively simple.

### Succint syntax

Under consideration is the possibility to create new pages in a succint manner:

```js
/pages/index/+Page.js
/pages/about/+Page.js
/pages/jobs/+Page.js
```

Here, a new `+config.js` file isn't needed each time a new page is created.

I'm currently leaning towards *not* supporting such succint syntax (in favor of "Single Config File", see next section). But I may reconsider if users want this.

### Single Config File (aka Single Route File)

```ts
// /pages/+config.ts

import type { Config } from 'vite-plugin-ssr'

const marketingPagesConfig = {
  Layout: './layouts/LayoutMarketingPages.vue',
  ssr: true,
}
const adminPagesConfig = {
  title: 'Admin Panel',
  Layout: './layouts/LayoutAdminPages.vue',
  ssr: false
}

export default {
  // If `singleConfigFile: true` then only one `+` file is allowed (this file). If there is
  // anothoer `+` file, then VPS displays a warning.
  singleConfigFile: true,
  pages: [
    { ...marketingPagesConfig, Page: './LandingPage.vue', route: '/'        , title: 'Awesome Startup'                          },
    { ...marketingPagesConfig, Page:    './JobsPage.vue', route: '/jobs'    , title: "We're hiring!"                            },
    { ...marketingPagesConfig, Page:  './VisionPage.vue', route: '/vision'  , title: 'Our mission <3'                           },
    {     ...adminPagesConfig, Page:   './AdminPage.vue', route: '/admin'   , onBeforeRender:   './AdminPage-onBeforeRender.ts' },
    {     ...adminPagesConfig, Page: './AdminDbPage.vue', route: '/admin/db', onBeforeRender: './AdminDbPage-onBeforeRender.ts' }
  ]
} satisfies Config
```

The `singleConfigFile` enforces the entire app to be defined in that single `+config.js` file. This means that this single `+config.js` file represents the entire interface between VPS and your app.

### Nested Layouts

```ts
// /pages/product/+config.js

export default {
  Page: './Page.js',
  onBeforeRender: './onBeforeRender.js',
  route: '/product/@id',
  nested: [
    {
      route: '/product/@id/details',
      View: './DetailsView.js'
    },
    {
      route: '/product/@id/reviews',
      View: './ReviewsView.js',
      onBeforeRender: './onBeforeRenderReviewsView.js'
    }
  ]
}
```

### Custom Exports

The V1 design requires [Custom Exports](https://vite-plugin-ssr.com/exports) to be registered.

```ts
// /pages/+config.ts

import type { Config } from 'vite-plugin-ssr'

export default {
  exportables: [
    {
      name: 'title',
      env: 'SERVER_ONLY' // | 'CLIENT_ONLY' | 'CLIENT_AND_SERVER'
    }
  ]
} satisfies Config
```

```ts
// /pages/index/+title.ts

// The file is loaded only in Node.js. (At server run-time or at build-time while pre-rendering.)
export const title = 'Welcome to the V1 design'
```

### More

The V1 design unlocks many more capabilities, in particular around building frameworks on top of VPS.

This is a very exciting design. (In many subtle ways, you'll see when you'll use a VPS framework.)

### Feedback

(Negative) feedback welcome.

**Comment by samuelstroschein** [comment]
Treat the following as notes, not explicit feedback or requests.

### singleConfigFile contradicts the nested example? 

- Are nested +config files auto-imported? 
- Why is `/pages/+config.ts` a single file config when `/pages/product/+config.ts` is an additional config file? Are the examples related? If not, how are nested layouts achieved with a single config file? 

### nested layout API via an array

using an array instead of a string for the `Layout` prop seems to enable very simple and intuitive nested layouts:


```js
const adminPagesConfig = {
  title: 'Admin Panel',
  // the first layout is the outer layout
  Layout: [
     './layouts/RootLayout.vue',
     './layouts/LayoutAdminPages.vue',
     // renders 
     //   <RootLayout>
     //      <LayoutAdmingPages>
     //           {children}
     //       </LayoutAdminPages>
     //    </RootLayout>
  ]
}

```

nested pages that define additional layouts can be merged easily with `parent.Layout.concat(child.Layout)`: 

```ts
// parent
export default {
  // additional parameters
  Layout: ["./Layouts/RootLayout.tsx"]
}

// renders ["./Layouts/RootLayout.tsx"]
```

```ts
// page1
export default {
  // additional parameters
  Layout: [
       "./Layouts/Product.tsx", 
       "./Layouts/Info.tsx"
   ]
}

// renders parent.Layout.concat(page1.Layout) = 
// ["./Layouts/RootLayout.tsx", "./Layouts/Product.tsx", "./Layouts/Info.tsx"]
```

```ts
// page2 is a child of page1
export default {
  // additional parameters
  Layout: ["./Layouts/Sidenav.tsx"]
}

// renders parent.Layout.concat(page1.Layout).concat(page2.Layout) = 
// ["./Layouts/RootLayout.tsx", "./Layouts/Product.tsx", "./Layouts/Info.tsx", "./Layouts/Sidenav.tsx"]
```


### the nested layout example seems unintuitive 

Defining the layout for nested pages in a page config seems unintuitive. Either each page is responsible for defining its layouts, or one config that defines everything. 

(I assume that `/product/@id/details` is a page too (if not why not?))

```ts
export default {
  // ...
  nested: [
    // why are layouts defined for sub-routes in a page config? 
    {
      route: '/product/@id/details',
      View: './DetailsView.js'
    },
  ]
}
```

**Comment by brillout** [comment]
@samuelstroschein

1. If you use `singleConfigFile` then you define `nested` inside your single config file `/pages/+config.js`:
   ```js
   // /pages/+config.js

   export default {
     pages: [
       {
         route: '/product/@id',
         Page: './product/ProductPage.js',
         nested: [
           {
             route: '/product/@id/details',
             View: './product/DetailsView.js'
           }
         ]
       }
     ]
   }
   ```
   Otherwise you define it in the page's config file `/pages/product/+config.js`:
   ```js
   // /pages/product/+config.js

   export default {
     route: '/product/@id',
     Page: './ProductPage.js',
     nested: [
       {
         route: '/product/@id/details',
         View: './DetailsView.js'
       }
     ]
   }
   ```
   Both are equivalent.
1. > Are nested +config files auto-imported?

    All file paths are auto-imported (VPS knows in what environment and when to import them).
1. Nested Views/Layouts are about parallelizing data fetching and that's why it needs to be built into the framework. Your proposed solution can already be achieved with VPS `0.4`.
1. Nested Views (note the naming "Nested Views" instead of Nested Layouts) aren't pages and don't behave like them. I believe the way Next.js, Nuxt and Remix does it is wrong. (Btw. nested views/layouts always have dedicated routes, see https://vite-plugin-ssr.com/layouts#nested-layouts.)

I've updated the RFC to better reflect this.

**Comment by ArnaudBarre** [comment]
I'm not doing any SSR right now so I can't give feedback on that side. Just taking the opportunity to suggest another syntax for config:

```ts
import type { Config } from 'vite-plugin-ssr'

export const config: Config = {
  // ...
}
``` 

This could avoid a lot of interop issues I think.

**Comment by brillout** [comment]
@ArnaudBarre 👍 From a TypeScript/user perspective I believe that would work equally well. I think it mostly boils down to a preference of having `export const config` vs `export default`. (It's reasonable to ask users to update their TS version. As for ESM/CJS compat it doesn't make a difference since VPS is always the only consumer of `+config.js` files.)

**Comment by astahmer** [comment]
regarding the `ssr` boolean option, where does that leave the current html-only (`x.page.server.ts` with no `x.page.ts`) rendering ? 

so it would completely replace & deprecate the current `{name}.page.{type}.ts` convention ? 
fwiw the `{name}.page.{type}.ts` felt intuitive to me, maybe others too since it's quite common in current meta-framework (Rakkas/Next/Remix/Nuxt/etc)
I suppose that the benefits are the same as Remix/tanstack router = knowing every route to avoid a fetching-cascade ?

anyway, it looks great, especially the custom exports ! congratz

**Comment by brillout** [comment]
@astahmer

HTML-only:

```ts
// +config.ts

import type { Config } from 'vite-plugin-ssr'

export default {
  configDefinitions: [
    {
      name: 'htmlOnly',
      handler(value) {
        if (typeof value !== 'boolean') throw new Error('Config htmlOnly should be a boolean')
        if (value) {
           return {
             name: 'Page',
             // Akin to .page.server.js
             env: 'SERVER_ONLY'
           }
        } else {
           return {
             name: 'Page',
             // Akin to .page.js
             env: 'CLIENT_AND_SERVER'
           }
        }
    }
  ]
} satisfies Config
```

```ts
// /pages/about/+config.js

export default {
  Page: './Page.jsx',
  htmlOnly: true
}
```
```ts
// /pages/about/Page.jsx

// This file is loaded only on the server

export default function() {
  return <>I'm a HTML-only page.</>
}
```

FYI, the idea long term is to have frameworks define such configs.

For example a VPS user is currently implementing a framework with island architecture. With V1 it works like this:

```ts
// node_modules/some-framework/+config.ts

import type { Config } from 'vite-plugin-ssr'

export default {
  configDefinitions: [{
    name: 'island',
    handler(value) {
      if (typeof value !== 'boolean') throw new Error('Config island should be a boolean')
      if (value) {
         return {
           clientEntries: ['./client-island-hydration.ts']
         }
      } else {
         return {
           clientEntries: ['./client-classic-hydration.ts']
         }
      }
    }
  }]
} satisfies Config
```

```ts
// /pages/about/+config.js

export default {
  Page: './Page.jsx',
  island: true
}
```
```ts
// /pages/about/Page.jsx

export default function() {
  return <>This page is island-hydrated</>
}
```

```ts
// /pages/+config.js

// Default value for all pages
export default {
  island: false
}
```

> fwiw the `{name}.page.{type}.ts` felt intuitive to me

One thing I don't like about V1 is that it's not clear anymore where files are being loaded. The user has to know it. Someone reading VPS user-land code for the first time will have no clue. A workaroud is to implement IDE plugins that visually display where a file is being loaded. (Btw. contributions welcome to implement prototypes.) Also, I see it to be the framework's responsability to clearly communicate where files are being loaded. Frameworks that are well designed and simple, together with such IDE plugins, solves the problem, I believe.

> the benefits

The foundational benefit here is that VPS knows everything about the user's app at build-time while enabling frameworks/users comprehensive control. The `configDefinitions` is the most powerful part of this RFC.

For the longest time it was believed that Next.js is the right level of abstraction. This RFC puts a big question mark on that assumption. Imagine a framework that deeply integrates with GraphQL.

```ts
// /pages/todo/+config.ts

import type { Config } from 'awesome-framework'

export default {
  query: '{ todo { text } }',
  Page: './TodoListPage.tsx',
  ssr: false // Not needed for a to-do list page
} satisfies Config
```

```ts
// /pages/todo/TodoListPage.tsx

// Ideally `awesome-framework` includes type generation
import type { Page } from './types'

export default function({ result }) {
  return <>{result[0].text}</> // displays "Buy milk"
} satisfies Page
```

Data fetching is completely taken care of by `awesome-framework`.

**Comment by astahmer** [comment]
makes sense ! thanks for the detailed explanations

**Comment by AaronBeaudoin** [comment]
I **love** this. 🤩

That doesn't mean I don't also have some questions/thoughts to add, however.

---

#### In your "islands" example above, how did `node_modules/some-framework/+config.ts` get loaded?

I suspect this question is a result of my lack of understanding, but how did VPS 1.0 find the `+config.ts` file at `node_modules/some-framework/`? I assume it doesn't search the entire `node_modules` directory for `+config.js` files, since that sounds like a performance nightmare. So then, how does it work? Am I correct in understanding that the `package.json` files look like this:

```js
// node_modules/some-framework/package.json
{
  ...
  "version": "0.1.0",
  "dependencies": {
    "vite-plugin-ssr": "^1.0.0"
  }
}

// ./package.json
{
  ...
  "dependencies": {
    "some-framework": "^0.1.0"
  }
}
```

And not like this:

```js
// node_modules/some-framework/package.json
{
  ...
  "version": "0.1.0"
}

// ./package.json
{
  ...
  "dependencies": {
    "vite-plugin-ssr": "^1.0.0", // Explicitly adding VPS as a dependency.
    "some-framework": "^0.1.0"
  }
}
```

Sorry if this is a bit of a dumb question!

---

#### I'd like to use my custom `configDefinitions` in my page files.

The `configDefinitions` idea is wild. I absolutely love it, because it lets me create a convenient "custom meta-framework API" for whatever I need for my project.

My question, however, is whether I can set my `configDefinitions` properties from my page files. In your "islands" example above (which blew my mind) you still set the `island` property in `pages/about/+config.js`. I'd like to just do `export const island = true;` from `pages/about/Page.jsx`. To explain why I would like to have this option, keep reading.

---

#### I'd also like to have a way to automatically have files be declared as a page.

The one "step backwards" in this new VPS architecture is the lack of any way for me to create some file `pages/whatever/SomePage.jsx` and have it automatically be picked up as a page without also having to add it in my `pages/+config.js` or `pages/whatever/+config.js`.

The thing is, although we now can have a "single" config file for VPS, it doesn't look like we can have a "set it once and never touch it again" config file. Every time I create a page, delete a page, modify a filename, or update anything about page's config (change it from server-only to client-only, or update a route) I need to head into my `+config.js` file. It is nice that this makes everything explicit, but it's also more time spent in configuration than I would like.

Maybe I could hack around this by having some custom file watching mechanism which updates my `+config.js` dynamically, but this seems like something I should be able to do a bit more easily, since VPS is already looking for some files automatically. 

### Here's my solution.

Give me a built-in `pagesFilesystemPattern` (or something like that) to tell VPS some set of files which should be interpreted as pages in the filesystem and automatically declared. It might look something like this:

```js
{
  ...
  pagesFilesystemPattern: /.+Page\.jsx/g
}
```

In this case any files ending in `Page.jsx` are automatically discovered and declared as pages by VPS. If I want to further customize my pages config I can still create `+config.js` files anywhere in my `pages/` subdirectories, in which case all pages at or below the current level will use the "overridden" config, but not require me to create `pages` properties in my `+config.js` files.

To have a bit more flexibility, I could also do something like this:

```js
{
  ...
  pagesFilesystemPattern: {
    pattern: /(.+?)(Server|Client)?Page\.jsx/g,

    // Not if this would be the arguments exactly, but you get the idea.
    handler(path, match) {

      // Return value is an object to append to `pages`.
      return {
        Page: path + "/" + match[0],
        route: path + "/" + match[1].toLowerCase(),
        env: {
          "Server": "SERVER_ONLY",
          "Client": "CLIENT_ONLY"
        }[match[2]]
      };
    }
  }
}
```

I could even use this to create a simple but powerful compatibility layer to migrate an existing VPS project to 1.0, such that all my existing `<whatever>.page.jsx` files get picked up automatically.

---

The two ideas above combined result in the ability for me to define a `+config.js` file once and never touch it again. I can give myself some way to create "filesystem" page files like is currently possible, but with my own custom convention, and then I can also give myself a way to affect the `+config` from inside my page file by doing some `export const <whatever>`.

I think this would be **beautiful**. I can basically have my own framework API from start to finish. If I've created a `node_modules/some-framework/+config.js`, then technically my own project can even have **zero** config.

**Comment by brillout** [comment]
@AaronBeaudoin

> how did `node_modules/some-framework/+config.ts` get loaded?

By using VPS's Extensions API. There aren't any docs for it yet, but you can have a look at an example [here](https://github.com/brillout/stem-react/blob/main/vite-plugin-ssr.config.ts).


> Am I correct in understanding that the `package.json` files look like this

Yes, `some-framework` can hide the VPS dependency away from the user.

> I'd like to just do `export const island = true;` from `pages/about/Page.jsx`.

That's not (easily) possible because configs (e.g. `ssr` or `island`) need to be known *before* transpilation begins, whereas `Page.jsx` can be loaded only *after* transpilation is done. The thing here is that `+config.js` is loaded before transpilation starts. (That's why it merely contains pointers to the actual application code.)

> with my own custom convention

While I share the sentiment to give further control to the user, I'd rather go for "convention over configuration" for this one. At the end of the day, I don't see the added value of allowing users to configure this. But I'm open to change my mind if I can see concrete benefits.


> The one "step backwards" in this new VPS architecture is the lack of any way for me to create some file `pages/whatever/SomePage.jsx` and have it automatically be picked up as a page without also having to add it in my `pages/+config.js` or `pages/whatever/+config.js`.

True.


I'm considering the possibility to create new pages in a succint manner:

```js
/pages/index/+Page.js
/pages/about/+Page.js
/pages/jobs/+Page.js
```

Here, a new `+config.js` file isn't needed each time a new page is created.

But I'm currently leaning towards *not* supporting such succint syntax. But I may reconsider if users want this.

**Comment by brillout** [comment]
@AaronBeaudoin

I'm thinking of this:

```js
// pages/+config.js

export default {
  // Tell VPS to also glob `+Page.js`
  glob: ['Page']
}
```

```jsx
// pages/about/+Page.jsx

export default () => <>This page was created without creating/touching any +config.js file</>
```

I think it's a good trade-off.

**Comment by npacucci** [comment]
Hello everybody! I would like to give my feedback about this, and ask two questions.

I'm actually using VPS to build a custom framework based on islands architecture (as you have already mentioned), so every page is server-side rendered and partial hydrated client-side.

For me it's great to have a clear distinction between `xx.server.tsx` and `xx.client.tsx` code, in terms of DX and robustness (considering also dependencies, .css/.scss imports and informations that needs to be passed from server to client granularly).

Anyway, also describing this within a config file can be ok, and even more powerful for other scenarios.

----------------
### First Question
First doubt is the same mentioned [here](https://github.com/brillout/vite-plugin-ssr/issues/53#issuecomment-1335727432). My cases are actually like:

 _test.page.server.tsx_:
```
// pages/test/test.page.server.tsx

export { passToClient };
export { onBeforeRender };
export { render };

const passToClient = [ ... "stuff to serialize" ... ];
const onBeforeRender = () => { ... "init code" ... }
const render = (pageContext) => { ... "full HTML page construction" ... }
```

plus  _test.page.client.tsx_:
```
// pages/test/test.page.client.tsx

export { render };

const render = (pageContext) => { ... "hydration code" ... }
```

And finally a custom route _test.page.route.ts_:
```
// pages/test/test.page.route.ts

export default '/test';
```

Reading your example I'm not sure if I got how this is gonna work with V1 purposed solution? Maybe like:

```
// pages/test/+config.ts

import type { Config } from 'vite-plugin-ssr'

export default {
  route: '/test',
  onBeforeRender: './test.page.before-render.server.ts', (Don't know if allowed name, should I put my "init code" here?)
  onRenderHtml: './test.page.render.server.tsx', // (Should I put my "full HTML page construction" code here?
  onRenderClient: './test.page.render.client.tsx', // (Should I put my  "hydration code" here?)
  ssr: true,
  Page: null, //(I don't need this, isn't it?)
  "stuff to serialize" here?? // (Should I list the actual passToClient strings?)
} satisfies Config
```

And even if I use a global +config, I assume what should I put inside it's the same as above, but for each page.

--------------------------
### Second Question

Will be possible to read pages handlers as external packages from _node_modules_? Like:
```
// pages/test/+config.ts

import type { Config } from 'vite-plugin-ssr';

export default {
  route: '@test-page/route.ts',
  onBeforeRender:'@test-page/test.page.before-render.server.ts',
  onRenderHtml: '@test-page/test.page.render.server.ts',
  onRenderClient:'@test-page/test.page.render.client.ts',
  ssr: true,
} satisfies Config
```

Or even more, read the full page `+config` as external package and declare it somewhere else (maybe into the "Single Config File"?

**Comment by brillout** [comment]
@npacucci

```ts
// pages/test/+config.ts

import type { Config } from 'vite-plugin-ssr'

export default {
  route: '/test',
  onBeforeRender: './onBeforeRenderTest.ts', // "init code"
  onRenderHtml: './onRenderHtmlTest.ts', // "full HTML page construction"
  onRenderClient: './onRenderClientTest.ts', // "hydration code"
  ssr: true,
  Page: './Page.ts', // You *do* need this: you can only omit it for the root pages/+config.ts
  passToClient: ['pageProps']
} satisfies Config
```

After talking with the Munich team, it seems to me that the RFC is actually quite a good match with BurdaForward's use case. Feel free to PM on discord so we can elaborate. Also, FYI, a VPS sponsor is currently building a framework VPS + Solid + Partial Hydration (ideally as a re-usable rendered like https://github.com/brillout/stem-react / https://github.com/brillout/vite-plugin-ssr/tree/main/examples/stem-react).

> Will be possible to read pages handlers as external packages from node_modules? Like:

Yes, that's actually a primary goal of VPS V1.

**Comment by brillout** [comment]
@npacucci

I forgot to reply about the following.

> For me it's great to have a clear distinction between `xx.server.tsx` and `xx.client.tsx` code, in terms of DX and robustness

I agree, my previous comment about this:

> One thing I don't like about V1 is that it's not clear anymore where files are being loaded. The user has to know it. Someone reading VPS user-land code for the first time will have no clue. A workaroud is to implement IDE plugins that visually display where a file is being loaded. (Btw. contributions welcome to implement prototypes.) Also, I see it to be the framework's responsability to clearly communicate where files are being loaded. Frameworks that are well designed and simple, together with such IDE plugins, solves the problem, I believe.

> And even if I use a global +config, I assume what should I put inside it's the same as above, but for each page.

Correct.

> Or even more, read the full page `+config` as external package and declare it somewhere else (maybe into the "Single Config File"?

I'm not entirely sure what you mean here, but you'll be able to publish npm packages that include `+config.js` files . (It's called the "Extensions API", you can see an example of the Extensions API with VPS 0.4 in the Stem React package I linked in my previous reply.)

**Comment by npacucci** [comment]
Great! I got it now. 

My only doubt is still about this:
`Page: './Page.ts', // You *do* need this: you can only omit it for the root pages/+config.ts`

Why this is needed? Cause even right now we are not using the [Page](https://vite-plugin-ssr.com/Page-server) feature, but only the [Render() hook](https://vite-plugin-ssr.com/render)

My prev example about this:
> // pages/test/test.page.server.tsx
> 
> export { passToClient };
> export { onBeforeRender };
> export { render };
> 
> const passToClient = [ ... "stuff to serialize" ... ];
> const onBeforeRender = () => { ... "init code" ... }
> const render = (pageContext) => { ... "full HTML page construction" ... }

So I expect to only set `onRenderHtml: './onRenderHtmlTest.ts',` (That should be the same of the actual Render() hook, isn't it?).

Otherwise if I have to set both `onRenderHtml` and `Page`, what should I write inside this last?

**Comment by brillout** [comment]
@npacucci Got it. So, yes, you can omit `Page` if you define both `onRenderHtml()` and `onRenderClient()`. (These two hooks are functionally equivalent to the current two hooks `.page.server.js#render` and `.page.client.js#render`.)

**Comment by brillout** [comment]
Example of using the V1 design: functionalities.

Note that V1 is still work-in-progess and functionalities beyond this basic example isn't supported yet.

**Comment by leonmondria** [comment]
Nice, I like the fact that it's less magic (some might call it convention), by configuring it this way it's more predictable and readable. 

I do think it's weird to see the + in the config file, never seen that before. Might scare some people, but might also just be something to get used to.

Small note, is there a reason why you are defining the imports with strings instead of direct imports?
```
import { onRenderHtml } from './config/onRenderHtml.tsx';
export default {
  onRenderHtml: onRenderHtml,
  // vs
  onRenderClient: './config/onRenderClient.tsx',
}
```

**Comment by brillout** [comment]
> weird to see the +

Yea, I found it ugly at first but it grew on me and I really like it now. (FYI it's SvelteKit who introduced that idea.)

> why [...] strings instead of direct imports?

Because the closure behavior would be confusing. Many criticize this about Next.js's `getServerSideProps()`, rightfully so.

```js
// /pages/+config.js

let app;

export default {
  onRenderClient(pageContext) {
    app; // This `app` variable is another than below
  },
  onRenderHtml(pageContext) {
    app; // This `app` variable is another then above
  }
}
```

**Comment by leonmondria** [comment]
> Because the closure behavior would be confusing. Many criticize this about Next.js's `getServerSideProps()`, rightfully so.

I agree, in that example it might lead to confusion. Still think it would be less magic/more predictable by moving the responsibility of actually importing, to the config file.

**Comment by brillout** [comment]
@vchirikov I will.

@leonmondria I share the sentiment here but I'm afraid directly importing in `+config.js` is quite problematic, for many reasons.

**Comment by iMrDJAi** [comment]
@brillout Here is a quick question: Is VPS going to continue supporting the current design?

Note: I find the `+` in `+config.js` too ugly, is the prefix gonna be customizable?

**Comment by brillout** [comment]
> Is VPS going to continue supporting the current design?

The `0.4` branch will be supporting both for a little while and soft-deprecate the old design (show a warning instread of throwing an error). The V1 release will completely remove the old design.

> Note: I find the `+` in `+config.js` too ugly, is the prefix gonna be customizable?

I found it ugly as well at first. But I grew to like it and I'm fairly confident you will as well. So no customization planned. As long as there are no functional regression then I think it's worth it to add a little bit of ugliness (in exchange for profound improvements, and also I've actually come to find it beautiful.)

**Comment by samuelstroschein** [comment]
Just here to express that we'd also like some glob matching of routes a la `*.page.tsx`. 

https://github.com/brillout/vite-plugin-ssr/issues/578#issuecomment-1369874203 proposes the following syntax:

```ts
// pages/+config.js

export default {
  // Tell VPS to also glob `+Page.js`
  glob: ['Page']
}

```

I suggest making the glob explicit. There seems to be no benefit of implicitly matching js/ts files with `+` prefix. Let users decide what to match explicitly:

```ts
export default {
  glob: ['*.page.{jsx|tsx}']
}
```

**Comment by brillout** [comment]
@samuelstroschein I'll think about it. (FYI there is a tension between the pro of having flexibility and the con of having too many options a la webpack. If they don't limit flexibiilty, then strong conventions, including naming conventions, are beneficial in the long term.)

**Comment by brillout** [comment]
I'm actually increasingly thinking of completely removing Filesystem Routing, including the aforementioned `glob` proposal. Single Route Files really are superior for anything non-trivial.

I understand that some folks want to build frameworks on top of VPS and want control over this. So I'm open to re-consider, but I'd like to engage into discussions: if you want Filesystem Routing then add a comment over there https://github.com/brillout/vite-plugin-ssr/discussions/630.

**Comment by brillout** [comment]
I've an idea how to make FS routing work very well with the V1 design. I'll show an example. We can consider this dicussion #630 as "closed" in the meantime.

**Comment by harrytran998** [comment]
Hi @brillout, do you had decided on what the V1 design looks like(final decision) and the roadmap for this? Currently, I start to migrate our projects from Nextjs to VPS but seem this process might take a long time 😆.

**Comment by brillout** [comment]
Example of the new Filesystem Routing: [examples/react-full-v1/](https://github.com/brillout/vite-plugin-ssr/tree/main/examples/react-full-v1). (E.g. see the `+` files in [renderer/](https://github.com/brillout/vite-plugin-ssr/tree/main/examples/react-full-v1/renderer) and [pages/hello/](https://github.com/brillout/vite-plugin-ssr/tree/main/examples/react-full-v1/pages/hello).)

@XiNiHa @AaronBeaudoin Ability to create custom configurations with deeper control than currently possible (the naming is messy but it *will* change and improve, I've already a couple of neat naming ideas): https://github.com/brillout/vite-plugin-ssr/blob/a9057be9beb08854034b5cc24572edf4292f1d02/examples/react-full-v1/renderer/+config.ts.

@harrytran998 The V1 is functionally equivalent to the current design (while allowing deeper control and customization than currently possible). So I recommend to start migrating to the current design and eventually migrate to the V1 design. The V1 design and the current design will co-exist for a while, so you'll be able to migrate on your own terms.

**Comment by samuelstroschein** [comment]
All the `+` files are too much boilerplate. 

I created an alternative proposal https://github.com/brillout/vite-plugin-ssr/pull/674

**Comment by redbar0n** [comment]
For inspiration on simplicity and API, I would postpone finalizing the V1 design until you could glance at how the upcoming TanStack Start framework is designed. Tanner has an impeccable taste for simple API's. https://tanstack.com/

**Comment by brillout** [comment]
@redbar0n What I like about what they're planning is typesafe routing and server code extraction (even though I disagree and I'd rather go with file based splitting as it's safer, see my reply in #674). But I don't think it's a reason for waiting on them. (FYI they don't have much yet &ndash; it's going to take a while until they ship something usable.)

**Comment by redbar0n** [comment]
I really don't like `Page.js` files, like in `/pages/about/+Page.js`, since they are confusing to discern from each other when multiple are open in IDE tabs (yes you can config IDE to show the path but most don't and it takes space).

Not sure if a `+` sign is even needed here, if vps could pick up on the file path or `.page.` in  the filename as the convention. (I generally don't like `+` in filenames, aesthetically, and struggle see why it's needed. Rails was able to do without it, based on convention, right?)

**Comment by brillout** [comment]
> discern from each other when multiple are open in IDE tabs

Yes, it's an issue I do consider. (I've a couple of ideas in mind, let's see.)

Curious: do you know whether VS Code can be configured to show the path?

> I generally don't like `+` in filenames, aesthetically

I guess it's a matter of taste. FYI I also didn't like it at first, but I grew to like it to the point that I now prefer it a lot more than `.page.`.

> why it's needed

The `+` prefix is about making it clear that it's is a VPS interface defining file. In general, I'd rather go for clear albeit more verbose than the opposite.

**Comment by redbar0n** [comment]
>Nested routes can overwrite parent +config files.
>>  VPS is configured via a root +config file with nested +config files per route.
>>> I'd love this, I really would. But I'm not sure how/if we can achieve this.

I support the option to have a general  / default config file, and an optional config file file near the components only if you need to override the default.

From https://github.com/brillout/vite-plugin-ssr/pull/674 :

> That's why with the current 0.4 design configs can live in 5 different places:
> 
>     .page.js
>     .page.client.js
>     .page.server.js
>     .page.route.js
>     vite.config.js (note that with the V1 design all configs previously defined in vite.config.js are now defined in +config.js)

I agree that config should live one place (for all pages), and optionally overridden at a single pre-determined other place (for each page). To not have to run around looking everywhere for the exact config.

Maybe the `.page.route.js` and `vite.config.js` for the page could be merged into a `.page.config.js`, since custom routing is a kind of config?
It seems you've already thought this way, considering the headline "Single Config File (aka Single Route File)".

Alternatively, for exceptional cases, any `.page.config.js` could be subdivided into (and fully replaced by):

```
.page.client.config.js
.page.server.config.js
```

(This subdivision might be useful In case you want to make some distinction between the client side routing and server side routing, or have a Route Function run on the client but not the server. Route Functions being a kind of config.)

The folder could alternatively also act as the page denotation instead of having `.page.` on every file. Convention over configuration. So it would be `/pages/error/error.config.js`, and if you view only the filename, without the path, you would just need to know from convention that any component with a config is a page.

Instead of:

```
/pages/error-page/error.page.client.config.js
```

Or the suggested:

```
/pages/error-page/+config.js
# or even
/pages/error-page/+isErrorPage.js
```

One could have:

```
/pages/error/error.client.config.js
```

Maybe one could even drop the /pages/ directory (like Next.js 13 does):

```
/app/error/error.client.config.js
```

(One would treat any component with `export { Page }` as a page, unless you need a `pagesFilesystemPattern`.)

In case of no separate config between server/client, it would simply be:

```
/app/error/error.config.js
```

Doesn't that look nice?

(That config file would include any potential Route Function for the `/app/error/error.js` page.)

Then `error.config.js` would also be much more distinguishable in IDE tabs, than multiple `+config.js` files.

---

So for normal pages, instead of:

```
/pages/index/+Page.js
/pages/about/+Page.js
/pages/jobs/+Page.js
```

You'd have:

```
/app/index/index.js
/app/about/about.js
/app/jobs/jobs.js
```

Alternatively, if you need a page distinction to target with regex in a `pagesFilesystemPattern`, then this could work, and be close to the existing 0.4 format:

```
/app/index/index.page.js
/app/about/about.page.js
/app/jobs/jobs.page.js
```

**Comment by redbar0n** [comment]
This might be a bit radical for a V1, but just to throw the idea out there:

You could even take it one step further, removing the "page" metaphor and allow any component to have/modify the URL. Could be very powerful, and a unique selling point for a framework. 

> "The page metaphor is killing the web." https://twitter.com/alexbunardzic/status/1633147283490627590?s=20 
> "The web is based on resources. Resources can be represented in countless ways. We chose to represent them as web pages. My hunch is that there's gotta be a better way to represent a resource on the web." https://twitter.com/alexbunardzic/status/1633161417607221248?s=20 
> "Sharing resource representations is the use case. Pages are just one way resources on the web can be represented." https://twitter.com/alexbunardzic/status/1633253175393423361?s=20
> 

Instead of making any component that `export { Page }` into a page, you could make any component that `export { Url }` affect the URL in some way.

 Or even better, check out: 
> “That means we can avoid hard-coding href attributes and create **reusable navigational components**.“ … “Navigational components contain Hyperlinks and Hyperlinks aren't reusable because their href attributes tie them down to a particular UI. The Navigation router's data-first approach sets Hyperlinks free so you can reuse them across UIs.” https://grahammendick.github.io/navigation/documentation/reusable-components.html

**Comment by redbar0n** [comment]
https://github.com/brillout/vite-plugin-ssr/issues/341 **Single Route File** contains a bit of information on the design of routing, maybe relevant to the discussion here.

**Comment by brillout** [comment]
> I agree that config should live one place (for all pages), and optionally overridden at a single pre-determined other place (for each page). To not have to run around looking everywhere for the exact config.

Yes and that's actually the case: everything is defined by `/pages/+config.js` (global) and `/pages/some-page/+config.js` (page-specific).

An expection being "shortcut" files such as:
 - `/pages/movie/+Page.js` being a shortcut for `/pages/about/+config.js > export default { Page }`
 - `/pages/movie/+onBeforeRender.js` being a shortcut for `/pages/about/+config.js > export default { onBeforeRender }`

Another exception is `/renderer/+config.js` which is equivalent to `/pages/+config.js`.

I understand the motivation for keeping VPS's interface minimal and I share that sentiment. But there are valid reasons to slightly bloat the interface with these two exceptions. (Although I'm thinking of removing `/renderer/+config.js` in favor of only supporting `/pages/+config.js`.)


> Maybe the `.page.route.js` and `vite.config.js` for the page could be merged into a `.page.config.js`

That's already the case.

```js
// /pages/movie/+config.js

export default {
  prerender: false,
  // This used to be defined in .page.route.js
  route: '/movie/@id'
}
```
```js
// /pages/+config.js

export default {
  // This used to be defined in vite.config.js
  prerender: true
}
```

Btw. you can already play with the V1 design: it's live on the `0.4` release branch. Simply copy an exampe with a `-v1` suffix.

> run on the client but not the server

With V1 you configure where code lives in `+config.js`:

https://github.com/brillout/vite-plugin-ssr/blob/6ae376ef74527514afef1e61455679d68e140425/examples/react-full-v1/renderer/+config.ts#L6-L10

You can even create a custom config like this:

https://github.com/brillout/vite-plugin-ssr/blob/6ae376ef74527514afef1e61455679d68e140425/examples/react-full-v1/renderer/+config.ts#L11-L29

:warning: The naming *will* change and improve. (E.g. I'm thinking of renaming `configDefinition` to `define` although it would conflict with Vite's [`define`](https://vitejs.dev/config/shared-options.html#define) so I'm not sure about the naming here.)

> any component with a config is a page.

That's already the case.

> `/pages/error/error.config.js`

To keep things simple I'd rather use a config: `/pages/error/+config.js > export default { isErrorPage: true }`.

> Maybe one could even drop the /pages/ directory [...] `/app/`

I find `pages/` 10x better/clearer than `app/`.

> `.config.js`

Regarding `.config.js` VS `+config.js`.

I prefer this:

```
/pages/movie/+config.js
/pages/movie/+Page.js
/pages/movie/+onBeforeRender.js
```

over this:

```
/pages/movie/index.config.js
/pages/movie/index.Page.js
/pages/movie/index.onBeforeRender.js
```

That's the main reason why I prefer `+config.js` over `.config.js`.

As for IDEs, honestly, I'm inclined to simply tell users to configure their IDE accordingly. It's possible for VS Code which is, by far, the most used one IDE. (I don't know its JavaScript market share but I guess it's growing and getting close to like 80%?)

> It [can be configured](https://stackoverflow.com/a/50181247/380607). But most won't do it.

If someone is annoyed by it but isn't able to change his VS code config, then we really can't help him :-).

> `pagesFilesystemPattern`

AFAICT it's a useless config. I'd rather find a good (enough) solution for everyone instead of giving the user too many useless options a la webpack.

> I hate special symbols with hidden meanings.

I don't see much of a difference between `/pages/movie/+config.js` and `/pages/movie/index.config.js`. I'd actually say that the former is easier to guess.


> "The web is based on resources. Resources can be represented in countless ways. We chose to represent them as web pages. My hunch is that there's gotta be a better way to represent a resource on the web."

In a future with digital interfaces la Minority Report, yes.

In the meantime, I think the page-centric web, as we know it today, works just fine. But I'm happy to be shown wrong with a *concrete* use case that adds enough added value to be worth exploring.

**Comment by redbar0n** [comment]
>> Maybe the .page.route.js and vite.config.js for the page could be merged into a .page.config.js.
> That's already the case.

Awesome. That's what I hoped. Thanks for the example.

>> any component with a config is a page.
> That's already the case.

Ok. But then why is a `.page.` or `Page` in the filename necessary to distinguish pages from other components?

> I prefer this:
> ```
> /pages/movie/+config.js
> /pages/movie/+Page.js
> /pages/movie/+onBeforeRender.js
> ```
> over this:
> ```
> /pages/movie/index.config.js
> /pages/movie/index.Page.js
> /pages/movie/index.onBeforeRender.js
> ```
> That's the main reason why I prefer +config.js over .config.js.

I agree, in that case there. But how about?

```
/pages/movie/config.js
/pages/movie/page.js
/pages/movie/onBeforeRender.js
```

Because why do you need the `+` if you can rely on filename convention?

Alternatively:

```
/pages/movie/movie.config.js
/pages/movie/movie.page.js
/pages/movie/movie.onBeforeRender.js
```

The file with the same name as the folder would be treated as the index. (I hate index files, they're indistinguishable.)

(Capitalized P in Page seemed out of place, so I lowercased it instead.)

**Comment by brillout** [comment]
> how about?
> 
> ```
> /pages/movie/config.js
> /pages/movie/page.js
> /pages/movie/onBeforeRender.js
> ```

The `+` is very much worth it, as it helps distinguish normal files from VPS files. It's asking too much from the user to remember all VPS's hooks and configs. Also, users (more realistically VPS frameworks) can create custom configs, so `+` is very much needed.

> Alternatively:
> 
> ```
> /pages/movie/movie.config.js
> /pages/movie/movie.page.js
> /pages/movie/movie.onBeforeRender.js
> ```

That'd be an alternative indeed. But I like the current implementation better because it's more compact:

```
/pages/movie/+config.js
/pages/movie/+Page.js
/pages/movie/+onBeforeRender.js
```

What could convince me to change my mind is this:

> As for IDEs, honestly, I'm inclined to simply tell users to configure their IDE accordingly. It's possible for VS Code which is, by far, the most used one IDE. (I don't know its JavaScript market share but I guess it's growing and getting close to like 80%?)

If someone can show me that enough IDEs with enough market shares cannot be configured, then I may reconsider the current design. Note that not many VPS users are complaining about this, so I'm inclined to think it isn't that big of deal.

P.S. the capital letter `P` in `+Page.js` should stay capital to keep things consistent. (It's exposed as `pageContext.Page` and not `pageContext.page`, which makes sense since it's a UI component and most UI frameworks have the naming convention to start UI components with a capital letter.)

**Comment by harrytran998** [comment]
I agreed with brillout about the file name configs. Besides, I think the focus of V1 is not the file name, it can be updated in the future when a lot of users complain about that(I think it does not happen, like Svelte Kit). 

If anyone comes to VPS, they care about the future, performance, and low-level architecture control. With the limited number of maintainers, I think DX should in `loose` mode.

**Comment by brillout** [comment]
> it can be updated in the future

Indeed. Using "soft deprecations" (support the old interface but show a warning telling users to move to the new interface) we can introduce changes in any minor release. The `1.0.0` release is mostly a marketing thing (it will be equivalent to the last `0.4.x` release minus the old interfaces). The V1 design is already released in `0.4.x`.

That said, I'm more than eager to hear (critical) feedback.

> I think DX should in `loose` mode.

I do care a lot about DX though :-).

**Comment by x6ax6b** [comment]
> P.S. the capital letter P in +Page.js should stay capital to keep things consistent.

Is a lowercase p (+page.js) an error? It's tricky if no proper lint (ex: unicorn/filename-case) is provided. It's just a convention for now.

**Comment by brillout** [comment]
@x6ax6b I believe https://github.com/brillout/vite-plugin-ssr/commit/334c303c15284ce539b486840895733683787179 addresses you concern.

@iMrDJAi Custom configs can have any name, built-in configs (should) have a unique name.

**Comment by spacedawwwg** [comment]
I really don't like the `+` at all. I agree with everybody else's sentiments on this.

Would 110% prefer this option over the what feels like super over opinionated file naming conventions.

```
/pages/movie/movie.config.js
/pages/movie/movie.page.js
/pages/movie/movie.onBeforeRender.js
```

If the `+` is a hill you're willing to die on @brillout (even though the feedback is pretty negative) can you not at least make the above an option for developers who want more verbosity?

"legibility trumps succinctness"

**Comment by spacedawwwg** [comment]
This is my current pages structure. This feel so much more organised and tidy. I'd hate to have to start implementing repetitive `+Page.ts` etc. I like being able to quickly scan my directory for file names.

<img width="276" alt="image" src="https://user-images.githubusercontent.com/1691867/225641396-98b292cc-ac67-47c1-9fed-fb34b0ddebdf.png">

**Comment by x6ax6b** [comment]
`+`+1
It's a special file and should be at the top of the file list in all cases.

**Comment by phiberber** [comment]
Just to mention. I've lost all my interest in SvelteKit due to these `+`s. I think they make the project structure feel cluttered, the same as other symbols as $ or _.

I do prefer the `.page.x` design, but I would rather use something like the new Next13 project structure with `page.vue`, `layout.vue`, `loading.vue`, etc. 

I think that replacing `.onBeforeRender.ts` with `.render.ts` and exporting a function `onBeforeRender` would be less confusing, even though these unused exports are not something I like, it's better in my opinion than big file names that also make it feel cluttered.

**Comment by brillout** [comment]
I can't speak for SvelteKit, but for VPS it seems pretty clear to me that the following is *much* clearer with than without `+`.

```
/pages/UcoHome/+Page.vue
/pages/UcoHome/+route.ts
/pages/UcoHome/+onBeforeRender.ts
/pages/UcoHome/someOtherNonVPSFile.ts
```

At the end of the day, it's only aesthetics. I'd rather go for ugly but clear, than "pretty" but confusing. (Personally, I actually now find `+` much prettier.)

I'm happy to be shown wrong with a *concrete* example why `+` is a problem.

As for your last point, you'll actually be able to only use `+config.ts` files:

```ts
// /pages/product/@id/+config.ts

import type { Config } from 'vite-plugin-ssr'

// No `+` needed here
import Page from './Page'
// (FYI, in case you're curious about the technicalities, VPS code-splits these imports,
// e.g. `onBeforeRender` is only loaded and executed on the server-side, while `Page`
// is loaded and executed on both the client- and server-side.)
import onBeforeRender from './onBeforeRender'

export default {
  Page,
  onBeforeRender
} satisfies Config
```

The capability of directly importing hooks in `+config.js` isn't implemented yet but it's on the roadmap.

**Comment by brillout** [comment]
> What if I import `beforeRender from './beforeRender.ts`?

So, yea, you'll be able to name your files and imports as you wish. So no more convention here. On the flip side, you'll have a complete overview simply by opening all `+config.js` files `$ git ls-files | grep +config | xargs vim`.

> Dangerous

VPS throws an error if you try to import `Page` and `onBeforeRender` from the same file (because VPS needs to be able to code-split per environment). In general, hooks can live in the same file only if they live in the same environment. It's well defined in what environments hooks are loaded, and VPS will throw a useful error accordingly.

**Comment by brillout** [comment]
That's your perspective which, as you describe it, seems quite subjective. As said, I'm happy to be contradicted but please with a rationale.

> I'm happy to be shown wrong with a _concrete_ example why `+` is a problem.

As I said many times here already, I also found `+` ugly *at first*. I can see that even *you* will eventually find it pretty.

**Comment by brillout** [comment]
The V1 design is now official released in `0.4.119`.

Migration: https://vite-plugin-ssr.com/migration/v1-design.

It's beta but I don't foresee breaking changes and some users have are already been using it.

Reach out if you have questions or run into any issue.

**Comment by seniorquico** [comment]
@brillout I scaffolded a new app using the latest (0.4.123) and began to work through a lot of the initial setup. I discovered this thread and "migrated" to the v1 design. All good, with one exception: Our app is quite large, and I setup the file structure based on the "domain-driven file structure" example. I ran into trouble trying to port the `_default.page.route.ts` file. No matter what I tried, I just kept getting 404's. See the comment I added below in the bit from [this example](https://vite-plugin-ssr.com/file-structure):

```
# Marketing pages
/marketing/pages/index/index.page.js # Landing page (URL: `/`)
/marketing/pages/about/index.page.js # About page (URL: `/about`)
/marketing/pages/jobs/index.page.js # Jobs page (URL: `/jobs`)
/marketing/_default.page.route.js # Route to `/*` instead of `/marketing/*` (see below)
# >> How do I correctly port ^^^ this route configuration to the v1 design? <<

# All code related to products
/products/pages/index/index.page.js
/products/pages/product/index.page.js
/products/pages/product/index.page.route.js
/products/db/fetchProduct.js
/products/db/fetchProductList.js
```

I tried to make a `+config.ts` file in the `/marketing` directory, among many other naive, wild guesses, but it couldn't seem to get it to work as I expected. Did I miss something obvious? Any guidance would be hugely appreciated. Also happy to contribute back a small comment to the v1 design migration doc if that's helpful.

**Comment by brillout** [comment]
@seniorquico Done & released in `0.4.126`. Domain-driven file structure (using `filesystemRoutingRoot`) migration example added at the end of https://vite-plugin-ssr.com/migration/v1-design#examples.

**Comment by brillout** [comment]
I recommend upgrading to `0.4.132`. Lot's of polishing around the V1 design has been done.

**Comment by louwers** [comment]
My 2 cents to the above discussion: unique filenames are very handy for finding the right file quickly. So I tend to avoid making lots of files with the same name.

**Comment by brillout** [comment]
> unique filenames are very handy for finding the right file quickly

What IDE are you using?

Personally I use https://github.com/junegunn/fzf (with Neovim) and it searches against the whole path (relative to the project root). Which means that, whether filenames are disambiguated doesn't make a difference. I'm assuming all IDEs would also match the search query against the full path and not only against the filename.

If you mean finding the right file among all open files and you use VSCode, then I'd suggest to configure VSCode to show the whole path instead of only the filename.

**Comment by louwers** [comment]
VSCode aggressively matches filenames before directories. So if you have a directory `profile/+Page.vue` and a file `program.js` somewhere it will put `profile.js` on top when you search for `pro`. If the file is falled `program-file.js` it will even match it first when you search for `prof`.

**Comment by marviobezerra** [comment]
The v1 specs are really nice, good job!

My only question, so far, is about the + sign. What is the purpose of this sign? What problem is it solving that a page.{jsx|tsx} wouldn’t solve?

NextJs is getting some negative feedback (including mine) on the amount of signs in files and folders - omg so many. In my opinion, simple is better, so no sign and file case insensitive or lower case.

**Comment by brillout** [comment]
@marviobezerra @DanAndreasson I do understand it's slightly off putting at first. But I believe you'll eventually end up liking it as it's a succinct way to clarify that the file is a vite-plugin-ssr entry.

> What problem is it solving that a page.{jsx|tsx} wouldn’t solve?

It doesn't solve a problem per se and an alternative such as `/pages/some-page/index.onBeforeRender.js` would work equally well. I explain the reason why I prefer `+` [here](https://github.com/brillout/vite-plugin-ssr/issues/578#issuecomment-1472064210).

> NextJs is getting some negative feedback (including mine) on the amount of signs in files and folders - omg so many. In my opinion, simple is better, so no sign and file case insensitive or lower case.

I agree with the sentiment and I also find Next.js's Filesytem Routing convention messy, and I'm confident vite-plugin-ssr won't turn into such a mess (a central reason being that vite-plugin-ssr allows you to define advanced routes programmatically, which means that vite-plugin-ssr can keep its Fillesystem Routing convention simple).

<br/>

To be clear: I'm not against revisting the `+` decision, but I'll ask you to make your case with rational arguments and not merely emotional ones. As said before multiple times, I do empathize that the `+` sign is off putting at first, and I also didn't like it at first. But I *did* end up liking it *a lot*, and I'm actually looking forward to deprecate the current convention in favor of the new `+` covention.

**Comment by marviobezerra** [comment]
I took a few days to answer because I was thinking about it (and also having some beers). There are many ways to approach a discussion like this one, and I agree with you that been rational is the best option. That been said, let get into it!

First, I'd chose convention over configuration 10 out of 10 times, specially for routing. Having a well defined file structure makes so much easier to navigate the code - specially for routes. In my option, the `+` sign in this case is "noise" and I'd avoid as much as I can. More over all my files are lower case, I do that so I can keep a smooth file system. When you have Pascal-case, Lower-case, Snake-Case, and Pick-Your-Poision-Case in the same application is like driving on a bumpy road.

Second, this kind of decision is tough because wherever you decide we need to cary on. I can imagine a developer creating a file `page.tsx` and not getting page rendered. After debugging and Googling for a few minutes (or more) they realize the it's missing the `+` and it should be `+page.tsx`. However, they still not getting the route to work. After more debugging and Googling they realize that it needs to be `+Page.tsx` (capital P). It may happens once or twice and the developer will learn its lesson (at least I hope so).

Also, when you jump between projects using different frameworks  you need to shift your brain to. For example, we have applications using NextJs, Expo, and React-Scripts (ugh). The `+` is one more wire to connect and it's not needed.

In summary, the `+` sign adds one small not necessary step that won't hurt and also won't help. As you said "It doesn't solve a problem per se" so it's a matter of liking or not. In the end it's your framework and it's your decision - be wise.

**Comment by brillout** [comment]
> the + sign adds one small not necessary step that won't hurt and also won't help

The + sign does have a function: it's about distinguishing normal JavaScript files from vite-plugin-ssr entries.

If we remove the + sign then:

```bash
# A vite-plugin-ssr entry
/pages/index/Page.js
# Another vite-plugin-ssr entry
/pages/index/onBeforeRender.js
# Confusing: is that a vite-plugin-ssr entry?
/pages/index/somethingElse.js
```

We need to have some kind of way to distinguish. So the question boils down to whether we should go for `/pages/product/+onBeforeRender.js` or something else, for example `/pages/product/index.onBeforeRender.js`.

> Also, when you jump between projects using different frameworks you need to shift your brain to. For example, we have applications using NextJs, Expo, and React-Scripts (ugh). The + is one more wire to connect when changing projects and, imo, it's not needed.

I don't see `/pages/product/index.onBeforeRender.js` to be better than `/pages/product/+onBeforeRender.js` in that regard.

> I can imagine a developer creating a file page.tsx and not getting page rendered. After debugging and Googling for a few minutes (or more) they realize [...] it should be +Page.tsx

You're right and I care about this. Actually the error message is quite nice already:

![c1](https://github.com/brillout/vite-plugin-ssr/assets/1005638/28f621ce-99af-4584-aa8f-98128617e412)

But I went ahead and made it better:

![c2](https://github.com/brillout/vite-plugin-ssr/assets/1005638/2c02d59b-a70f-4013-8c3d-fa343285d305)


> they realize the it's missing the +

Because `/pages/product/+onBeforeRender.js` is much more "in your face" than `/pages/product/index.onBeforeRender.js`, I'd actually argue that the user is much less likely to forget the + sign than forgetting the `index.` prefix.

> When you have Pascal-case, Lower-case, Snake-Case, and Pick-Your-Poision-Case in the same application is like driving on a bumpy road.

I actually appreciate that SvelteKit is going for snake_case over camelCase (I prefer snake_case a lot more) but, unfortunately, I think they're fighting a lost cause: the JavaScript ecosystem as a whole will most likely stick to camelCase.

I don't think there is an alternative here other than taking the bullet.

> More over, all my files are lower case, I do that so I can keep a smooth file system.

I'd actually recommend you to forgo that practice of yours because the JavaScript is pretty much ubiquitously camelCase.

> the first impression

I agree the first impression isn't perfect here. But, even with considering the first impression, I still prefer this:

```
/pages/product/+Page.vue
/pages/product/+route.ts
/pages/product/+onBeforeRender.ts
```

Over this:

```
/pages/product/index.Page.vue
/pages/product/index.route.ts
/pages/product/index.onBeforeRender.ts
```

I find the increased compactness appealing.

It find it so appealing that I think it's worth the inital impression of weirdness. (And after using it a while, users won't find it weird anymore.)

I'm open to alternatives to `/pages/product/+onBeforeRender.js` and `/pages/product/index.onBeforeRender.js`. I can't think of any but maybe someone can come up with one.

> I took a few days to answer because I was thinking about it (and also having some beers). There are many ways to approach a discussion like this one, and I agree with you that been rational is the best option. That been said, let get into it!

Thanks. I appreciate it.

**Comment by marviobezerra** [comment]
I see that you already made your mind. It'd amazing to have a simple solution, but, again, your framework and it's your decision.

```
/pages/product/{P|p}age.vue
/pages/product/{R|r}oute.ts
/pages/product/on{B|b}efore{R|r}ender.ts
```

Thanks for taking the time to answer.

**Comment by brillout** [comment]
> ```
> /pages/product/{P|p}age.vue
> /pages/product/{R|r}oute.ts
> /pages/product/on{B|b}efore{R|r}ender.ts
> ```

FYI Next.js had a lot of negative feedback because of the distinguishing issue of such design.

**Comment by phiberber** [comment]
Am I right to say that there's more negative feedback than positive on this topic? 

I know you think people are going to get used to it (I didn't yet), but I'd like to know the ratio of people enough to convince you emotionally.

It's a "Noise" vs "Useful Noise" battle which already looks in my opinion pretty personal and emotional.

A battle which I have no motivation to participate anymore.

**Comment by marviobezerra** [comment]
I’m also done.

**Comment by redbar0n** [comment]
Just throwing this suggestion out there:

/pages/product/product.ts
/pages/product/product.vike.ts

Where the product.ts file by convention would contain the index and solve the problem of multiple index.ts files.

Where the Vike (upcoming new name for VPS) file would contain all of:

- route config (only included in case it should override the default / general route config)
- Page component (capitalized export, like the convention, but no capitalized file name needed)
- onBeforeRender

Wouldn’t this be clean? Thumbs up if so.

**Comment by samuelstroschein** [comment]
The *.vike.* filename clearly communicates that "this is a vike related file" much better than implicit "+config files mean it's related to vike" 👍

**Comment by marviobezerra** [comment]
> That's precisely my point: I want to be convinced on a rational basis, not on a merely emotional one.

What is your motivation other the emotional? I agree that it needs reserved file names, so `+Page`, `page`, or `page.vike` won’t matter. Only your preference (aka emotional).

> I wouldn't be sure about it: it could very well be a small minority that is a lot more vocal than a vast agreeing majority.

Maybe a survey can provide actual data to support a decision. Formulate a valid question here and ask for thumbs up, thumbs down, etc. Would you accept the result?

**Comment by flodlc** [comment]
After many days testing the v1 design, I actually like the "+" syntax a lot 👍 

- "+" explicitly means that the file export will be added in the config file
- Readability and short syntax
- Easy to distingue Components from config files
- "+" at the start of the filename makes alphabetical order works without mixing with non "+" files, which is in my opinion not negotiable.

Previous design leads to lot of export in the page file and it became hard to see which one is present.

**Comment by brillout** [comment]
@redbar0n The V1 design requires each config to be defined by its own file (or in the `+config.js` file when the value is primitive/serializable). It's at the core of the V1 design and enables foundational features such as a config `ssr: boolean`. An `ssr` flag needs to change in which environment `+Page.js`/`Page.vike.js` is loaded.

That said, a `.vike.js` suffix would indeed be an alternative. And I like it.

That said, `+` is more succint and has the advantage of improved FS order as pointed by @flodlc. So, personally, I'm still in favor of the + sign.

I appreciate your feedback which is *much* more constructive than just saying "I categorically don't like it and I don't want to explain why".

> Only your preference (aka emotional).

It isn't. It's rational. For example, the + sign is the most succint of all suggestions.

> Maybe a survey [...] Would you accept the result?

I would consider the result but I wouldn't make it the sole reason for deciding. Also, I wouldn't do the survey now as it'd be premature and shortsighted: let's first see how many get used to the + sign after a couple of months. Remember, my argument is that most will get used to the weirdness. (Maybe it will become a widespread convention even outside of SvelteKit and Vike...). It's my job to see far in the future.

That said, I can make a survey out of curiosity.

**Comment by phiberber** [comment]
I've mentioned before that the reason I don't like the + sign is how cluttered it look, mentioned also an alternative that would be using a colon which in my opinion looks really simple and it's one of the things I did when I started my own framework.

> It isn't. It's rational. For example, the + sign is the most succinct of all suggestions.

I've suggested two options, the colon instead of the plus sign and the way Next does it (Although I don't like that much how colocation works in Next).

> That said, a .vike.js suffix would indeed be an alternative. And I like it.

I actually prefer the plus sign over a suffix, I'm looking for the less cluttered way to identify pages and a suffix is just too much to me.

> I would consider the result but I wouldn't make it the sole reason for deciding. Also, I wouldn't do the survey now as it'd be premature and shortsighted

I agree with you on that one and change my opinion on this topic, I still hate the plus sign but maybe you're right in your motivation to wait, but it's quite strange to implement something that you're thinking in possibly removing in the future.

If I've interpreted right, all the things @flodlc mentioned would be possible with a colon instead of a plus, I don't think there's a rationale to that, it's just less cluttered to my eyes and I agree it's extremely personal, emotional and subjective, you should do that too @brillout.

**Comment by redbar0n** [comment]
@phiberber 

> On Windows systems, files and directory names cannot be created with a colon (:). But if a file or directory name is created with a colon on a Linux or Mac operating system, then moved to a Windows system, percent encoding is used to include the colon in the name in the index.

https://www.ibm.com/docs/en/spectrum-archive-sde/2.4.1.0?topic=tips-file-name-characters#:~:text=On%20Windows%20systems%2C%20files%20and,the%20name%20in%20the%20index.

**Comment by phiberber** [comment]
Makes sense, I'm not a Windows user, Maybe - then, something cleaner than +.

**Comment by brillout** [comment]
@phiberber Seems like you're slowly appreciating the whole + thing after all :-).

Beyond redbar0n's point the issue with `:` is that it already denotes parameterized route segments in many other ecosystems.

> Maybe - then, something cleaner than +

I like the fact that `+` can be interepreted as "adding something" which kinda makes sense. The `-` sign is weird from that perspective.

**Comment by phiberber** [comment]
I still hate it and it has been one of the three reasons I avoided VPS. But I'm seeing there's not that many options of symbols other than +, suffixes are bad, prefixes are good but they should not feel cluttered, which is my main problem with the + sign.

Seeing a large project tree with + signs is the same to me as looking to every ad at the Times Square, too much information to process.

Colocation feels badly designed to me still, it could maybe get better with proper IDE support.

+Page.ts
$Page.ts
@Page.ts
>Page.ts
<Page.ts
_Page.ts
-Page.ts
.Page.ts
~Page.ts

To me - means more like an item of a list than subtraction or something negative, it's really clean and easier to read.

**Comment by brillout** [comment]
That's actually something I'd be up for making a poll about and going with the result. But still premature.

Out of all options, I clearly prefer `+`. So I'd suggest we go for `+` and revisit the sign after a while. I'm hopeful you'll get used to it.

**Comment by samuelstroschein** [comment]
What about `vike.page.tsx`, `vike.config.ts` to achieve grouping of vike related files and more explicitness than "`+` means it's related to vike"?

**Comment by brillout** [comment]
The + sign quickly ends up being as explicit as a `vike.` prefix. A `vike.` prefix is more explicit only for newcomers, and I don't think it justifies giving away the succinctness of the + sign.

I'd suggest we pause and resume the discussion after a couple of months after the old design is deprecated.

That said, I'm more than looking forward to hear about concrete issues, such as paper cuts around IDE. (I actually have an idea about it but I purposely don't write about it because, so far, I don't think it's enough of a priority for us to spend time on. Important to prioritize things correctly, otherwise Vike will never hit prime time.)

To sum up:
 - Let's go for the + sign and revisit after a couple of months.
 - If you stumble upon concrete issues, e.g. with your IDE, let us know about it.

Looking forward to ship the V1 (design).

**Comment by redbar0n** [comment]
@phiberber 

I do also like - more than + purely aesthetically. But I think + may be better here.

> To me - means more like an item of a list than subtraction or something negative, it's really clean and easier to read.

I agree, I never thought about it as subtraction and addition, in terms of file names. - is well established in file names, and should be intuitive to most.

But, related, an issue I have with YAML is that it’s confusing to quickly glance at something and tell whether it is a list or not. Because of - being present or not (esp. when nested).

That’s could actually be an issue here too, since it would look like some folder items in the IDE indicate they are part of a list, so the (inexperienced) user may be inclined to make all files in the folder be prefixed with - to be uniform. Or vice versa. Both may conflict with this library’s operation.

**Comment by Hebilicious** [comment]
Everything looks exciting !

I have one minor suggestion : 

```ts
export default {
  configDefinitions: [
  {  name: 'island', handler(value) {}},
  {  name: 'foo', handler(value) {}},
  ]
  } satisfies Config
  ```

Like this ...
 
 ```ts
export default defineVikeConfig({
  extendConfig: {
 	island: { handler: (value) => {  // object syntax }},
	foo : (value) => { // array syntax }
  }
 })
 ```

The reasons why  :
 
- Using a config helper is helpful for people who don't want to use Typescript as it provides type info for them as well.
- `extendConfig` is arguably semantically more correct than `configDefinitions`
- Using an object to define the config is shorter while remaining readable and extensible

**Comment by brillout** [comment]
@Hebilicious It's actually already renamed, see https://vike.dev/meta. I find "meta" better than "extendConfig", given the fact that `meta` is also about modifying existing configurations.

As for the config helper, it didn't occur to me that it would help JS users. Neat & good to know. PR welcome if that's something you'd be up to.

I also like the idea of an object syntax. Maybe something like that:

```js
// /pages/some-page/+config.h.js

import dataHook from './data.js'

export default defineConfig({
  // object syntax
  dataMeta: {
    value: dataHook,
    // By default `env` is 'server-only'
    env: 'server-and-client'
  },
  // direct syntax
  data: dataHook
})
```

```js
// /pages/some-page/data.js

export default (pageContext) => {
  /* ... */
}
```

It's a succinct way to define both the config value and its options :+1:.

**Comment by Hebilicious** [comment]
> @Hebilicious It's actually already renamed, see https://vike.dev/meta. I find "meta" better than "extendConfig", given the fact that `meta` is also about modifying existing configurations.
> 
> As for the config helper, it didn't occur to me that it would help JS users. Neat & good to know. PR welcome if that's something you'd be up to.
> 
> I also like the idea of an object syntax. Maybe something like that:
> 
> ```js
> // /pages/some-page/+config.h.js
> 
> import dataHook from './data.js'
> 
> export default defineConfig({
>   // object syntax
>   dataMeta: {
>     value: dataHook,
>     // By default `env` is 'server-only'
>     env: 'server-and-client'
>   },
>   // direct syntax
>   data: dataHook
> })
> ```
> 
> ```js
> // /pages/some-page/data.js
> 
> export default (pageContext) => {
>   /* ... */
> }
> ```
> 
> It's a succinct way to define both the config value and its options :+1:.
> 

I'll open a PR then !

About meta, it's more likely that people associate it with metadata, metaverse or Facebook meta ...

In what context is meta about modifying configuration ?

In my opinion, the name should encapsulate something like define, extend, modify or add.

**Comment by brillout** [comment]
I agree and I ain't 100% happy with the name "meta" but it's the best I could find so far. The use cases listed in https://vike.dev/meta are a good representation for what `meta` is used for. Contribution welcome to suggest a better name.

(Looking forward to your `defineConfig()` PR 👀)

**Comment by vchirikov** [comment]
@brillout it will be nice to have docs about v1 at vike.dev, because for now it's unlisted :\

**Comment by brillout** [comment]
@redbar0n I think, given the context, `metaConfig` would be redundent.

```js
// /pages/+config.js

export default {
  // Since we are in +config.js it's fairly clear that meta relates to "meta configuration".
  meta: {
    // ...
  }
}
```

<br/>

@vchirikov I agree and it will happen as soon as the V1 design is out of beta. (And since we increasingly recommend users to use Bati and `vike-{react,vue,solid}`, which both use the V1 design, we should release the V1 design sooner rather than later.)

There is one little breaking change coming for the V1 design and then it should be ready for release. So it's coming soon.

**Comment by redbar0n** [comment]
@brillout yeah I thought about that, but then again, since a config can refer to anything the meta can refer to anything not necessarily a type of config. So with just `meta` then it might seem that you are configuring some particular thing called `meta` (re: all kinds of associations with metadata, metaverse or Facebook meta ...) and not generally applying _a configuration of the configuration itself_.. Anyways, for my part I found risking a little over-specification in this case wouldn't hurt. Better too clear than unclear.

**Comment by brillout** [comment]
Technically, `meta` *is* a config, just like any other config. How a config is being used is completely up to the implementation that uses it (that's one of the beauty of the V1 design as it's very flexible/customizable). And, actually, `meta` isn't the only config that influences other configs, e.g. `ssr: boolean` changes the `env` of `Page`. (FYI this flexibility is paramount for `vike-{react,vue,solid}` and other upcoming integrations such as `vike-relay`.)

**Comment by Hebilicious** [comment]
> I agree and I ain't 100% happy with the name "meta" but it's the best I could find so far. The use cases listed in [vike.dev/meta](https://vike.dev/meta) are a good representation for what `meta` is used for. Contribution welcome to suggest a better name.
> 
> (Looking forward to your `defineConfig()` PR 👀)

For alternative names : 

- `extendMeta`
- `extendDefinitions`
- `extendConfig`
- `extendKeys`
- `extendMetaKeys`
- `extendConfigKeys`
- `extendConfigDefinitions`

As I said before, `extend(...)` can be replaced by `add`, `additional`, `define` ...

**Comment by redbar0n** [comment]
Re: https://vike.dev/meta

My vote goes to `customConfig` since it's:
- self-explaining and straightforward. Everyone has head about a "custom config" before.
- not an overloaded term like `meta`. Not everyone has heard about a "meta config" before.
- since it relates to the WHAT and not the HOW (create, update), it encapsulates both creation and update. It also makes sense since it is _declaring_ a JSON object to send to Vike. _Imperative_ names would make more sense for functions.
- gives a good feeling that you can actually customize Vike to your hearts content.
- implies that you'll rely on Vike's default config if you don't provide a custom one.

**Comment by brillout** [comment]
I actually did consider the name `define`, but I ended up going for `meta` as it does a better job at communicating that it's a config that influences configs, which is pretty much what "meta level" means. And `define` can mean everything really (it actually already means something for bundlers like Vite and webpack so, let alone for that reason, we cannot pick that name).

As for the other suggestion, they are misnomers as they convey either extending or modifying but not both. Same for `custom` (which I actually also did consider back then): it conveys creating new configs but it doesn't convey modifying existing configs. I understand that `customConfig` is slightly clearer but it's also clunky.

I think `meta` is good enough, so let me suggest we settle for it.

**Comment by redbar0n** [comment]
It’s intuitively understood by everyone that a custom config overrides a default config. It’s pretty much a standard. 

Sometimes it’s the things we think about at first that are the best solutions, because that will also be what others intuitively think first. :-)

**Comment by phiberber** [comment]
Tailwind had that problem with theme replacing the default config instead of modifying it, they sure had a lot of questions because people weren't used to the `extend` prefix.

Not sure how much the extend prefix could be useful here, although it doesn't seem to me that it's really a config, the current state of the meta parameter looks to me more like route props than a config itself.

If I'm not wrong with the above thought, I'd say meta is better than extendConfig or customConfig, although in that sense I think ``props`` would be a better name.

**Comment by redbar0n** [comment]
Precedence and inspiration:

https://remix.run/docs/en/main/file-conventions/remix-config#routes

> remix.config.js
>
>A function for defining custom routes, in addition to those already defined using the filesystem convention in app/routes. Both sets of routes will be merged.

**Comment by brillout** [comment]
The [V1 design is out of beta](https://twitter.com/brillout/status/1731654463352610957) and now the official design.
